wallet, rpc: add UTXO set check and incremental rescan to importdescriptors

musaHaruna opened the pull request:
Fixes [#28898](https://github.com/bitcoin/bitcoin/issues/28898)

When importing descriptors, users may accidentally provide an incorrect birthdate (timestamp). This can cause the wallet to miss relevant historical transactions, leading to incorrect or incomplete balances. Currently, the wallet only relies on rescans starting from the provided timestamp.

This PR extends the importdescriptors RPC with a new optional argument:

`verify_balance` a `bool` which is `false` by defualt.

If enabled, the wallet will compare its calculated trusted balance against UTXO set balance (by generating the `scriptpukeys` of the wallet and comparing it with the chains UTXO set `scriptpubkeys` to get the accurate balance belonging to the wallet and comparing it with the wallet trusted balance).

If the balances match, import continues as normal.
If a discrepancy is detected, the wallet will attempt incremental rescans in chunks of recent blocks until the missing history is found. If the wallet is pruned, incremental rescans will not go earlier than the prune boundary.

Additional information is returned in the RPC response under an `"info" ` object when `verify_balance` is used, they are:

`utxo_check`: whether the UTXO set matched the wallet balance
`scanned_chunks:` number of incremental rescan chunks attempted
`scanned_blocks`: approximate number of blocks scanned during incremental rescans

This helps detect and fix balance mismatches caused by wrong descriptor timestamps.

**Implementation:**
- Add `FindCoinsByScript()` for scanning UTXO set for coins belonging to output_scripts and expose it to node interface through `findCoinsByScript()`.
- Add `GetWalletUTXOSetBalance()` which calculates the wallet’s total spendable balance by summing all UTXOs in the blockchain’s UTXO set that belong to the wallet’s scriptPubKeys, ignoring immature coinbase outputs.
- Add `std::optional<int64_t> endTime` to `RescanFromTime()` which lets the rescan stop at a specific block height, limiting the scan to a defined time range instead of scanning from startTime all the way to the blockchain tip.
- Add `IncrementalRescan()` this function scans the wallet’s transaction history in chunks of blocks from tip backwards to try and reconcile the wallet’s trusted balance with the actual UTXO-set balance, stopping early if a chunk produces a balance that matches the target. It is used only when there’s a discrepancy between the wallet’s trusted balance and the UTXO-set balance, performing an incremental rescan to identify when the balances align and recording how many chunks were scanned. 

**`importescriptors` with incremental example**
```
bitcoin-cli importdescriptors '[{
    "desc": "wpkh([f5b1c3d7/84h/0h/0h]xpub6CUGRUonZSQ4TWtTMmzXdrXDtypWKiKpJrZT4S9vPz4ZZ6f4a1V4f3ChFy8hRk4M6mh4J4LR6WcJbNf2oyMFGmQH6Eo1QvLKMjR18q3fTq5/0/*)#abcd1234",
    "timestamp": "now"
}]' true
```

Example Output
```
[
  {
    "info": {
      "utxo_check": true,
      "scanned_blocks": 2000,
      "scanned_chunks": 2
    }
  }
]
```

**Benchmarks**
Here's a benchmark result for normal rescan and incremental rescan as suggested by **fjahr**. 

| ns/op        | op/s       | err% | total | benchmark                   |
|-------------:|-----------:|-----:|------:|:----------------------------|
| 13,481.56    | 74,175.41  | 0.9% | 0.01  | BenchmarkIncrementalRescans |
| 13,450.49    | 74,346.72  | 1.4% | 0.01  | BenchmarkIncrementalRescans |
| 13,468.04    | 74,249.85  | 1.5% | 0.01  | BenchmarkIncrementalRescans |
| 2,274,875.00 | 439.58     | 1.7% | 0.03  | BenchmarkRescanFull         |
| 2,301,375.00 | 434.52     | 1.4% | 0.03  | BenchmarkRescanFull         |
| 2,306,334.00 | 433.59     | 3.0% | 0.03  | BenchmarkRescanFull         |


Here's the [code](https://github.com/bitcoin/bitcoin/commit/15982b5cdef97d89c5ac68fe4b075abad1f65228)  for the benchmark. Not 100 percent sure if I wrote it correctly. It's my first time writting benchmark, took inspiration from how [src/bench/wallet_balance.cpp](https://github.com/bitcoin/bitcoin/blob/master/src/bench/wallet_balance.cpp) was written.


A big thanks to **fjahr** for suggesting this approach [comment](https://github.com/bitcoin/bitcoin/pull/33392#issuecomment-3300319224)

musaHaruna committed:
node: add and implement `FindCoinsByScript()` for UTXO-set scan by script

Add `FindCoinsByScript()` which implements a  DB-backed scan
that iterates the on-disk UTXO DB with a `CCoinViewCursor` and wire it into the node
interface as `Node::getCoinsByScript()`.

musaHaruna commented:
>  Have you looked into making this an option of importdescriptors? 
I have not looked into making it an option for `importdecriptors`, but I will look into it and check the performance, and get back to you on that.  

Thanks for the suggestion. I really appreciate it.

musaHaruna commented:
> I have not looked into making it an option for importdecriptors, but I will look into it, and get back to you on that.

Yes, I have looked into adding the option to importdescriptors by introducing a scan_utxoset flag and it's possible. The idea is that when enabled, the wallet would scan the UTXO set immediately after import to verify balances against chainstate, and if a discrepancy is detected (for example due to an incorrect birthdate), it could automatically trigger a full rescan from height 0 to restore missing history. This way, users wouldn’t have to manually diagnose incomplete balances — the import flow itself would handle it. What Do you think this approach?.

I don’t yet know the full impact on performance, but I’m thinking of making the trade-offs very clear in the RPC docs so users can decide whether to enable it.

musaHaruna commented:
> Maybe run some benchmarks on the utxo set scan and full rescan and add them to the PR here. I don't have a good feeling for what the relation is and this might influence what reviewers think is better concerning running a full rescan automatically or just returning a warning/hint from `importdescriptors`.

I will run some benchmarks on both utxo set scan and full rescan, update the whole PR to use the new approach you suggested i.e to add the feature on `importdescriptors` directly, which I honestly think is better. 


> Another idea concerning performance: It seems likely that if the birth date is wrong the user might be off by just a few days or weeks, rather than they have a wallet with the satoshi coins. So instead of a full rescan from the start of the chain the rescan could move backwards from the originally supplied birth date and scan the chain in 1000 (or so) block increments and it could stop once the balance matches the one from the utxo set scan. A nice side effect of this is that it means the rescan is also pruning compatible (for as many blocks that are available) but I guess this could be also achieved by starting from pruneheight instead of 0 if pruning is enabled. It's something you need to keep in mind either way. This would also need benchmarks but I can't imagine moving backwards with somewhat large increments would make the process much slower. And ideally the process would exit early in the most common scenario.

Just to make sure we’re on the same page — my understanding is that the idea is to:

Use chunked backward rescans starting from the supplied birthdate (e.g. 1000-block increments), and stop as soon as the wallet balance matches the UTXO-set scan. This handles the common case where the birthdate is only off by a few days or weeks, so we don’t need to rescan the entire chain.

At the same time, respect the pruneheight as the lower bound if pruning is enabled. That means we only scan back as far as blocks are available, and if the discrepancy still isn’t resolved at that point, we’d warn the user that a reindex is required.

fjahr commented:
> Just to make sure we’re on the same page — my understanding is that the idea is to:
> 
> Use chunked backward rescans starting from the supplied birthdate (e.g. 1000-block increments), and stop as soon as the wallet balance matches the UTXO-set scan. This handles the common case where the birthdate is only off by a few days or weeks, so we don’t need to rescan the entire chain.

Right, so for example if the supplied birthdate was blockheight `800,000` but there is a discrepancy, then, instead rescanning from 0 to 800k right away, scan 799k - 800k and check the balance, exit if it's a match, continue with 798k - 799k and so on. This shouldn't be too complex to implement since the `rescanblockchain` RPC already takes a start and an end height.

> At the same time, respect the pruneheight as the lower bound if pruning is enabled. That means we only scan back as far as blocks are available, and if the discrepancy still isn’t resolved at that point, we’d warn the user that a reindex is required.

Yeah, there would be some kind of safe abort if the approach for walking backwards runs into unavailable blocks, just like `rescanblockchain` does as well.

luke-jr commented on src/wallet/rpc/coins.cpp:
```diff
@@ -160,22 +213,36 @@ RPCHelpMan getreceivedbylabel()
                     {"minconf", RPCArg::Type::NUM, RPCArg::Default{0}, "Only include transactions confirmed at least this many times."},
                     {"include_watchonly", RPCArg::Type::BOOL, RPCArg::Default{false}, "No longer used"},
                     {"avoid_reuse", RPCArg::Type::BOOL, RPCArg::Default{true}, "(only available if avoid_reuse wallet flag is set) Do not include balance in dirty outputs; addresses are considered dirty if they have previously been used in a transaction."},
+                    {"scan_utxoset", RPCArg::Type::BOOL, RPCArg::Default{false}, "If true, scan the UTXO set and return scanned UTXO balance alongside the trusted balance."},
```
This shouldn't be a positional parameter, at least.

luke-jr reviewed (commented):
I agree getbalance* is the wrong place for this kind of check.

But it also will fail to detect incorrect birthdates if the TXOs are spent, so it can't be relied on either...

musaHaruna committed:
wallet: add `GetWalletUTXOSetBalance()` to calculate balance from UTXO set

Introduce a new helper `GetWalletUTXOSetBalance(const CWallet&)` that
derives the wallet’s balance directly from the UTXO set rather than from
the wallet’s transaction history.

Usefull for verifying wallet balance correctness against the
chainstate without requiring a full rescan.

musaHaruna committed:
wallet/rpc: add `verify_balance` arg & docs for importdescriptors

Add a new optional boolean RPC argument `verify_balance` to
importdescriptors and document the response fields used when the
`verify_balance` check is enabled.

musaHaruna committed:
wallet: extend` RescanFromTime()` with optional endTime to limit rescan range

Add an optional endTime parameter to `CWallet::RescanFromTime()`.
When provided, the method finds the approximate end height and calls
`ScanForWalletTransactions()` with a bounded [start, end] range.
This makes incremental rescans possible while preserving existing
behavior when no endTime is specified.

rkrux reviewed (commented):
Concept ACK and cursory review 1ed236e3c686582e35f20ab668cd6d93316fae2c.

The intent of the PR and the corresponding issue #28898 seems fine to me.

musaHaruna commented on src/wallet/rpc/backup.cpp:
```diff
@@ -401,10 +499,73 @@ RPCHelpMan importdescriptors()
+
+    UniValue utxo_diff_obj;
+
+    if (do_scan_utxoset && rescan) {
```
The UTXO set check is performed before the main rescan to attempt incremental reconciliation of missing funds in small block chunks, potentially reducing the amount of blocks that need to be rescanned in the case of a large blockchain. Doing it after the main rescan would generally be redundant, because the rescan will already incorporate transactions missed due to incorrect timestamps, and any discrepancy with the UTXO set would likely have been resolved. The current order allows us to detect and fix large balance mismatches earlier in the RPC, without necessarily scanning the entire chain from genesis immediately.

musaHaruna commented on src/wallet/rpc/backup.cpp:
```diff
@@ -323,6 +323,7 @@ RPCHelpMan importdescriptors()
                             },
                         },
                         RPCArgOptions{.oneline_description="requests"}},
+                        {"scan_utxoset", RPCArg::Type::BOOL, RPCArg::Default{false}, "If true, scans the UTXO set balance and compare with wallet balance and triggers incremental rescan if discrepency is found."}
```
Yeah I agree I have changed it to ` verify_balance` I thought just `verify` is till a bit vague

musaHaruna commented on src/wallet/rpc/backup.cpp:
```diff
@@ -401,10 +499,73 @@ RPCHelpMan importdescriptors()
+    bool have_prune_boundary = false;
+    int64_t min_trial_start_time = 0;
+
+
```
Fixed

musaHaruna commented on src/wallet/rpc/backup.cpp:
```diff
@@ -401,10 +499,73 @@ RPCHelpMan importdescriptors()
+        UniValue early = IncrementalRescansNonOverlap(wallet, tip_time, tip_height, chunk_blocks, avg_block_time, have_prune_boundary,
+                                                        min_trial_start_time, utxo_scanned_balance, reserver, response, out_chunks_tried, out_lowest_ts);
+
+        if (!early.isNull()) {
```
Yeah, I think it should be, I will look into how I can do that, am open to suggestions on how that can be done.

musaHaruna committed:
rpc: extend importdescriptors with UTXO check and incremental rescan

Extend the importdescriptors RPC with an optional `scan_utxoset` parameter.
When enabled, the wallet balance is compared against a balance derived from
the UTXO set. If a discrepancy is found, the wallet attempts incremental,
non-overlapping rescans in fixed-size block chunks until the balances match,
allowing for faster recovery in many cases. When a match is detected, the
RPC returns early with an `info` object describing the UTXO check and rescan
progress. If no match is found, the code falls back to the existing full
rescan behavior starting from the earliest descriptor timestamp (or the
prune boundary if applicable). Responses are updated to include optional
`info` fields with `utxo_check`, `scanned_chunks`, and `scanned_blocks`
metadata.

musaHaruna committed:
test: add importdescriptors `verify_balance` flag behavior

musaHaruna reviewed (commented):
Resolved conflict with [#34049](https://github.com/bitcoin/bitcoin/pull/34049).
Addressed:

> scanned_blocks is computed as chunks_scanned * 1000 here, but BLOCKS_PER_CHUNK is defined as a local constexpr inside IncrementalRescan and isn't accessible there. Consider either moving BLOCKS_PER_CHUNK to file scope, or adding a blocks_scanned field to IncrementalRescanResult. That way the computation stays inside IncrementalRescan.

By adding a blocks_scanned field to IncrementalRescanResult

Thanks for the review @rxbryan.

junbyjun1238 commented on src/wallet/rpc/backup.cpp:
```diff
@@ -36,6 +37,92 @@
+    // Get current chain tip
+    int64_t tip_time{0};
+    int tip_height{0};
+    LOCK(wallet.cs_wallet);
```
The new `verify_balance` paths hold `cs_wallet` across heavy scan operations where the existing `importdescriptors` flow releases it before rescanning.

`GetWalletUTXOSetBalance()` takes `LOCK(wallet.cs_wallet)` while collecting scriptPubKeys, but keeps the lock held through `findCoinsByScript()`, which iterates the entire on-disk UTXO set. `IncrementalRescan()` takes `LOCK(wallet.cs_wallet)` before the loop and never releases it, so the entire chunk-rescan loop, including repeated `RescanFromTime()` and `GetBalance()` calls, runs under the wallet lock.

Consider scoping the lock to just the scriptPubKey collection / tip lookup, then running the UTXO-set scan and chunk rescans without it. As written, `verify_balance=true` can block wallet RPC/GUI operations that need `cs_wallet` for the duration of those scans, potentially hundreds of 1000-block chunks on mainnet.

musaHaruna reviewed (commented):
Thanks @junbyjun1238  for the review. 

Apologies for the delayed response — I got carried away working on [js-bitcoinkernel](https://github.com/musaHaruna/js-bitcoinkernel).

I've addressed [this](https://github.com/bitcoin/bitcoin/pull/33392#discussion_r3278350771) by scoping `cs_wallet` to only the wallet data access. In `GetWalletUTXOSetBalance()`, the lock is now held only while collecting the wallet's scriptPubKeys and is released before `findCoinsByScript()`. Similarly, in `IncrementalRescan()`, the lock is now scoped to retrieving the wallet tip `(GetLastBlockHash()/findBlock())`, and the incremental rescan loop runs without holding `cs_wallet`, allowing other wallet operations to proceed during the long-running scan. Changes are in [ca63e2d](https://github.com/bitcoin/bitcoin/pull/33392/commits/ca63e2d41db581385d4d0aba09aa54297a7dbca4) and [fceb84e](https://github.com/bitcoin/bitcoin/pull/33392/commits/fceb84e1970d26d464d3a2b0e9fe8cb57e630eb6)

musaHaruna reviewed (commented):
Rebased and fix conflict with [336f5a7](https://github.com/bitcoin/bitcoin/pull/35579/changes/336f5a738b3d89abb8dd7ebdc4b31ac12a6de85f)

haanhvu commented on src/wallet/rpc/backup.cpp:
```diff
@@ -422,12 +526,40 @@ RPCMethod importdescriptors()
+        if (utxo_scanned_balance != wallet_balance.m_mine_trusted) {
+            incremental_rescan_result = IncrementalRescan(wallet, utxo_scanned_balance, reserver);
+
+            if (incremental_rescan_result) {
```
In the other case (when `incremental_rescan_result` is `nullopt`), should we give any info to users, at least something like we tried incremental scanning but failed to get results? Otherwise they may get confused why they set `verify_balance = true` but don't get any result/info?

haanhvu commented on src/wallet/rpc/backup.cpp:
```diff
@@ -476,8 +608,19 @@ RPCMethod importdescriptors()
                 }
             }
         }
+    } else {
+        std::vector<UniValue> results = response.getValues();
+        response.clear();
+        response.setArray();
+
+        if (incremental_rescan_result->matched) {
+            for (unsigned int i = 0; i < results.size(); ++i) {
+                UniValue res = results[i];
+                res.pushKV("info", info);
+                response.push_back(res);
+            }
+        }
```
Response is cleared before `if (incremental_rescan_result->matched)` is checked. So if incremental rescan finishes with the wallet's balance unmatching the UTXO-set balance, we'll have empty response?

Is that what we really want? If not, should we:

```suggestion
    } else {
        if (incremental_rescan_result->matched) {
            std::vector<UniValue> results = response.getValues();
            response.clear();
            response.setArray();
            
            for (unsigned int i = 0; i < results.size(); ++i) {
                UniValue res = results[i];
                res.pushKV("info", info);
                response.push_back(res);
            }
        }
```
?

haanhvu commented on src/wallet/rpc/backup.cpp:
```diff
@@ -37,6 +38,94 @@
+    while (true) {
+        result.chunks_scanned++;
+
+        result.blocks_scanned += BLOCKS_PER_CHUNK;
```
I think we can get the real number of blocks scanned per chunk here instead of approximating it with the hardcoded `BLOCKS_PER_CHUNK` (1000). `RescanFromTime()` calls `ScanForWalletTransactions()`, and in `ScanResult` returned from `ScanForWalletTransactions()` we already have the `last_scanned_height`:
https://github.com/bitcoin/bitcoin/blob/58a7869f860f022e6163dc3fae3106a0599830c5/src/wallet/wallet.h#L640-L654
Should we do something like `last_scanned_height - start_height + 1` to calculate the real number of blocks scanned per chunk?

Another way is explicitly incrementally counting every time a block's scanned in `ScanForWalletTransactions()`:
https://github.com/bitcoin/bitcoin/blob/4dacd7c9da4529ec8f8203e4c53ea11c361e5d30/src/wallet/wallet.cpp#L1980-L1982

haanhvu reviewed (commented):
Approach ACK
